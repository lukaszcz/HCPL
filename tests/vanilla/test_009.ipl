
// eagerness & laziness test

let g = \&x { print x; g };

let f0 = \&x \!y {
   if == x 0 then
      y
   else {
      print 1;
      f0 (- x 1) (y 0)
   }
};

let f1 = \&x \&y {
   if == x 0 then
      y
   else {
      print 1;
      f1 (- x 1) (y 0)
   }
};

let h = \&x { print 7; print x; + x x; };
let he = \!x { print 7; print x; + x x; };

f0 5 g;
print 7;
f1 5 g;
print (h {print 2; 3});
print (he {print 2; 3});
print (he &{print 2; 3});

let! x = {print 8; 9};
print (h x);

let& y = {print 8; 9};
print (h y);

print (h {print 8; 9});

let aux0 = \!x {
   print 7;
   print x;
};

let aux1 = \!x {
   print 7;
   aux0 x;
};

aux1 &(&{print 8; 4});

let delay = \n \&x {
   print 7;
   if == 0 n then x else delay (- n 1) &(&x)
};

let force = \n \&x {
   print 1;
   if == 0 n then x else force (- n 1) !x
};

print (force 11 (delay 5 {print 3; 2}));
