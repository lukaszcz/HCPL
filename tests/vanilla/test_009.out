1
0
1
0
1
0
1
0
1
0
7
1
1
1
1
1
0
0
0
0
0
7
2
3
6
2
7
3
6
7
2
3
6
8
7
9
18
7
8
9
18
7
8
9
18
7
7
8
4
1
7
7
7
7
7
7
1
1
1
1
1
1
1
1
1
1
3
1
2
